content/post/2019-10-07-Image-Creation-Process.md

```diff
@@ -0,0 +1,53 @@
+---
+date: 2019-10-07T00:00:00-00:00
+description: ""
+featuredimage: ""
+tags: []
+title: "The Image Creation Process"
+show_reading_time: true
+authors:
+- name: cloudtoad
+  link: https://twitter.com/cloudtoad
+---
+
+Each lesson in NRELabs includes a sandbox environment for learners to interact with so they can explore lesson concepts on real systems.  This environment consists of one or more docker images.  An image could be a linux host, a virtual network appliance, an IP PBX, or really anything that a lesson creator can dream up.  But how do these images get created?  In this post, we’ll walk through the three stages of image creation.
+
+## Stage 1: Building a Base, Ex Nihilo
+
+Image creation starts with building a base.  This is done by downloading an install ISO for a linux distribution, or a prepared image such as a virtual network appliance.  
+
+We boot this image to add the “antidote” user and add it to the appropriate user groups.  In most cases we add this user to the “sudo” group, or otherwise give this user equivalent administrative privileges.  For instance, in the Cumulus VX image, we add the user to the “netshow” and “netedit” groups so it can run Cumulus’s “net” tool.  Next, we ensure that the primary interface of the image is configured for DHCP.  
+
+In the case of applications like Asterisk PBX, Free-Range Routing, or Netbox images, we install or compile the relevant packages and dependencies.  On all images, we ensure that any APIs (including REST or Netconf) are enabled.
+
+Once we are done with the ISO image itself, we need to create a dockerfile for the image that launches it with QEMU.  The launch script for this container should look for and create the right network interfaces with QEMU, and stitch them to the underlying interfaces in the container.  It should also pass a mount-point to the virtual-machine so it can mount the lesson and configuration files.  Finally, the launch script should ensure the right ports are being forwarded from the underlying primary interface of the container (“eth0”) to the primary interface of the virtual-machine.
+
+The completed ISO files for QEMU are stored in a Google Cloud Platform bucket.  The docker images are stored in the “antidotelabs” docker hub repository.  Our base is done!
```

Comment:
This paragraph makes it sound like google cloud buckets are a build destination, and we're trying to only store unmodified disk images there. Maybe rephrase to mention that we store raw unmodified images (and are striving to do that for everything) in google cloud, but the only destination for this pipeline is docker hub - that would be more accurate.

Reply:
I'll update the verbiage.